fish-rust/src/wildcard.rs

```diff
@@ -11,3 +48,1249 @@ pub const ANY_STRING_RECURSIVE: char = char_offset(WILDCARD_RESERVED_BASE, 2);
+    // base dir is /, and the wildcard is the remainder. If we don't find one, the prefix is the
+    // working directory, the base dir is empty.
+    let (prefix, base_dir, effective_wc) = if wc.starts_with(L!("/")) {
+        (L!(""), L!("/"), wc.slice_from(1))
+    } else {
+        (working_directory, L!(""), wc)
+    };
+
+    let mut expander = WildCardExpander::new(prefix, flags, &mut cancel_checker, output);
+    expander.expand(base_dir, effective_wc, base_dir);
+    return expander.status_code();
+}
+
+/// Test whether the given wildcard matches the string. Does not perform any I/O.
+///
+/// \param str The string to test
+/// \param wc The wildcard to test against
+/// \param leading_dots_fail_to_match if set, strings with leading dots are assumed to be hidden
+/// files and are not matched (default was false)
+///
+/// \return true if the wildcard matched
+#[must_use]
+pub fn wildcard_match(
+    name: impl AsRef<wstr>,
+    pattern: impl AsRef<wstr>,
+    leading_dots_fail_to_match: bool,
+) -> bool {
+    let name = name.as_ref();
+    let pattern = pattern.as_ref();
+    // Hackish fix for issue #270. Prevent wildcards from matching . or .., but we must still allow
+    // literal matches.
+    if leading_dots_fail_to_match && (name == L!(".") || name == L!("..")) {
+        // The string is '.' or '..' so the only possible match is an exact match.
+        return name == pattern;
+    }
+
+    // Near Linear implementation as proposed here https://research.swtch.com/glob.
+    let mut px = 0;
+    let mut nx = 0;
+    let mut next_px = 0;
+    let mut next_nx = 0;
+
+    while px < pattern.len() || nx < name.len() {
+        if px < pattern.len() {
+            match pattern.char_at(px) {
+                ANY_STRING | ANY_STRING_RECURSIVE => {
+                    // Ignore hidden file
+                    if leading_dots_fail_to_match && nx == 0 && name.char_at(0) == '.' {
+                        return false;
+                    }
+
+                    // Common case of * at the end. In that case we can early out since we know it will
+                    // match.
+                    if px == pattern.len() - 1 {
+                        return true;
+                    }
+
+                    // Try to match at nx.
+                    // If that doesn't work out, restart at nx+1 next.
+                    next_px = px;
+                    next_nx = nx + 1;
+                    px += 1;
+                    continue;
+                }
+                ANY_CHAR => {
+                    if nx < name.len() {
+                        if nx == 0 && name.char_at(nx) == '.' {
+                            return false;
+                        }
+
+                        px += 1;
+                        nx += 1;
+                        continue;
+                    }
+                }
+                c => {
+                    // ordinary char
+                    if nx < name.len() && name.char_at(nx) == c {
```

Comment:
ok so we know that given "c == ANY_CHAR && nx == name.len()" we will never match this condition,
because name never contains ANY_CHAR. Therefore it's fine to include that case in the above ANY_CHAR

Reply:
In the C++ code, the case of `c == ANY_CHAR && nx = name.len()` was handled in the "ordinary char case" not in the `ANY_CHAR` case, that's what was interesting